main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala

```diff
@@ -561,45 +562,40 @@ object GenExpression {
 
       case AtomicOp.Is(sym) =>
         val List(exp) = exps
-        // We get the JvmType of the class for tag
-        val classType = JvmOps.getTagClassType(sym)
-        // First we compile the `exp`
+        val taggedType = BackendObjType.Tagged
+
         compileExpr(exp)
-        // We check if the enum is `instanceof` the class
-        mv.visitTypeInsn(INSTANCEOF, classType.name.toInternalName)
+        val ins = {
+          import BytecodeInstructions._
+          CHECKCAST(taggedType.jvmName) ~ GETFIELD(taggedType.NameField) ~
+            pushString(JvmOps.getTagName(sym)) ~ INVOKEVIRTUAL(JavaObject.EqualsMethod)
+        }
+        ins(new BytecodeInstructions.F(mv))
 
-      // Normal Tag
       case AtomicOp.Tag(sym) =>
         val List(exp) = exps
-        // We get the JvmType of the class for tag
-        val classType = JvmOps.getTagClassType(sym)
-        // Find the tag
-        val tag = root.enums(sym.enumSym).cases(sym)
-        // Creating a new instance of the class
-        mv.visitTypeInsn(NEW, classType.name.toInternalName)
-        mv.visitInsn(DUP)
-        // Evaluating the single argument of the class constructor
-        compileExpr(exp)
-        // Descriptor of the constructor
-        val constructorDescriptor = AsmOps.getMethodDescriptor(List(JvmOps.getErasedJvmType(tag.tpe)), JvmType.Void)
-        // Calling the constructor of the class
-        mv.visitMethodInsn(INVOKESPECIAL, classType.name.toInternalName, "<init>", constructorDescriptor, false)
 
-      case AtomicOp.Untag(sym) =>
+        val tagType = BackendObjType.Tag(BackendType.toErasedBackendType(exp.tpe))
+
+        val ins = {
+          import BytecodeInstructions._
+          NEW(tagType.jvmName) ~ DUP() ~ INVOKESPECIAL(tagType.Constructor) ~
+            DUP() ~ pushString(JvmOps.getTagName(sym)) ~ PUTFIELD(tagType.NameField) ~
```

Comment:
Neat to call String.intern

Reply:
Ahh, need to call String.intern, i see 👍